### PR TITLE
Improve documentation and code style

### PR DESCRIPTION
**Pull request by bancer**

### CODE REVIEW

File: lib/Cake/Network/CakeResponse.php [hunk 1]
```diff
@@ -854,7 +855,7 @@ public function sharable($public = null, $time = null) {
 		}
 
 		$this->maxAge($time);
-		if (!$time) {
+		if ($time === null) {
```
Comment (dereuromark) [inline lib/Cake/Network/CakeResponse.php]: What about 0 or false or other fasely values?
By changing this here you possibly introduce dangerous BC breaks for existing apps.
Better not changing functionality in a maintenance release major anymore where not absolutely necessary. Here it doesnt seem like it.

File: lib/Cake/Network/CakeResponse.php [hunk 2]
```diff
@@ -1464,7 +1465,7 @@ protected function _sendFile($file, $range) {
 		$file->open('rb');
 
 		$end = $start = false;
-		if ($range) {
+		if (!empty($range)) {
```
Comment (dereuromark) [inline lib/Cake/Network/CakeResponse.php]: the previous was 
a) more correct as it doesn't cloak
b) otherwise completely identical

please see cakephp coding guidlines for this.

File: lib/Cake/Network/CakeResponse.php [hunk 3]
```diff
@@ -717,7 +718,7 @@ public function httpCodes($code = null) {
  *
  * e.g `type(array('jpg' => 'text/plain'));`
  *
- * @param string $contentType Content type key.
+ * @param array|string $contentType Content type key.
```
Comment (dereuromark) [inline lib/Cake/Network/CakeResponse.php]: `|null` missing if you want to be correct/precise.

File: lib/Cake/Network/CakeResponse.php [hunk 4]
```diff
@@ -717,7 +718,7 @@ public function httpCodes($code = null) {
...
  * @return mixed current content type or false if supplied an invalid content type
```
Comment (dereuromark) [inline lib/Cake/Network/CakeResponse.php]: string|false then?

File: lib/Cake/Network/CakeResponse.php (hunk 1 above)
Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: > The expression $time of type integer|null is loosely compared to false; this is ambiguous if the integer can be 0. You might want to explicitly use === null instead.

The code intention was not very clear here. Perhaps `if (empty($time)) {` would be better.

Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: No, "if empty" is also bad (and functionally equal to what was before), see my other message.
`if (!$time)` is perfect to be kept.

File: lib/Cake/Network/CakeResponse.php (hunk 2 above)
Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: > The expression $range of type array is implicitly converted to a boolean; are you sure this is intended? If so, consider using ! empty($expr) instead to make it clear that you intend to check for an array without elements. 

According to CakePHP [coding standards](https://book.cakephp.org/2.0/en/contributing/cakephp-coding-conventions.html#comparison) regarding comparison:
> Always try to be as strict as possible.

File: lib/Cake/Network/CakeResponse.php (hunk 1 above)
Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: The intention of the check becomes ambiguous when an interger is converted to a boolean. Another option would be to cast to an integer `if ((int)$time === 0) {` or to set the default value to 0 instead of `null` what I think makes more sense as `$time` is supposed to be an integer: 
`public function sharable($public = null, $time = 0) {`.

`if (!$time) {` check is also inconsistent with strict check `if ($seconds !== null) {` done in `maxAge()` method that is called on line 857.

File: lib/Cake/Network/CakeResponse.php (hunk 2 above)
Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: You didnt fix strictness here, you just did the opposite. You cloaked a var, which is bad.
Read the part about not using empty() and isset() wrongly.

File: lib/Cake/Network/CakeResponse.php (hunk 1 above)
Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: If that is the case, then casting sure would be more BC here, people could rely on falsely values to also trigger the ! part.

File: lib/Cake/Network/CakeResponse.php (hunk 2 above)
Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: Could you be more specific in referencing to the guidelines? I have not found any mention of `empty` on that page. I do not fully understand how making the intention of the code clearer makes it worse.

Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: It is pretty highlighted on the page actually:
https://book.cakephp.org/3.0/en/contributing/cakephp-coding-conventions.html#careful-when-using-empty-isset

You may also read [this](https://www.dereuromark.de/2016/06/01/use-isset-and-empty-wisely/).

Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: > When working with variables, it is better to rely on type-coercion to boolean instead of empty().

That statement is highly arguable. In my opinion it is quite confusing and is in fact buggy when a string or an array is converted to a boolean for comparison that way. 

In this particular case `list()` that is called below expects `$range` to be an array. If `$range` is `true` or `1` (and this may happen if `true` or `1` is passed as the parameter of `_sendFile` method) this code makes no sense:
```
list($start, $end) = true;
```
The values of `$start` and `$end` are `null` in this case. 
```
$end = $start = false;
var_export($start); // outputs false
var_export($end);   // outputs false
list($start, $end) = true;
var_export($start); // outputs NULL
var_export($end);   // outputs NULL
```
And that seems to be not what is expected here. The code inside `if ($start !== false) {` below would not be executed.

So the correct check would be `if (is_array($range) && !empty($range)) {` here.

Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: Let's please stop that discussion.
You are wrong here in terms of functionality. It does not make a difference, those statements (before and after your change) are identical. The downside of cloaking is explained in both linked resources. This has nothing to do with casting.
Please revert this and focus on the actual changes that matter.

File: lib/Cake/Network/CakeResponse.php [hunk 5]
```diff
@@ -1464,7 +1465,7 @@ protected function _sendFile($file, $range) {
 		$file->open('rb');
 
 		$end = $start = false;
-		if ($range) {
+		if (is_array($range) && !empty($range)) {
```
Comment (dereuromark) [inline lib/Cake/Network/CakeResponse.php]: You are changing behavior again:
If the range is not an array, the whole functionality here differs now instead of blowing up as it should.

As discusses, please revert fully.

Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: Again: 2.x is in maintenance mode, security and critical bugfixes only.
No functional changes for the sake of changing anymore, please, to avoid further accidental regressions.

File: lib/Cake/Network/CakeResponse.php (hunk 2 above)
Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: I've got an impression that you did not fully read my previous post :(.

Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: I did - but you are then not understanding how PHP works. There is no casting happening.
The $range is still what it is, even after such a check. http://sandbox.onlinephpfunctions.com/code/580d6dbc977dd40112d36dcb8976662cc9d6e0bf

Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: I understand that no casting happens and the value of `$range` does not change. That's why I wrote "and this may happen if true or 1 is passed as the parameter of _sendFile method". The values of `$start` and `$end` can be null in some scenarios. http://sandbox.onlinephpfunctions.com/code/6623b50ebc865963924eb4fa8730c38bd2d4d922

Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: Is there a test case to proof this edge case?
Otherwise the change below looks too dangerous.

Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: If there would be a unit test for this scenario the bug would not be in the code. Why the code in http://sandbox.onlinephpfunctions.com/code/6623b50ebc865963924eb4fa8730c38bd2d4d922 is not convincing?

Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: Maybe [this](http://sandbox.onlinephpfunctions.com/code/989185d2a0747f4cf2dbc15714b567d1b3a5ccd1) would be more convincing?

File: lib/Cake/Network/CakeResponse.php (hunk 5 above)
Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: Use `if ($range && is_array($range)) {` then to fit the guidelines.

Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: You are still wrong about what it means: in PHP it doesnt mean true, but not empty in the context of arrays. Your code is cloaking which is 10x worse. Let's please stop this discussion.
The more important question is: Why the array check now? What happens if it is not an array?
Then it would silently continue doing sth different then it did before.

Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: > in PHP it doesnt mean true, but not empty in the context of arrays.

I think you are wrong about that. From PHP documentation:

> As described in the section about expressions, expression **is evaluated to its Boolean value**.
http://php.net/manual/en/control-structures.if.php

> When necessary, an expression is **automatically converted to boolean**.
http://php.net/manual/en/language.expressions.php

> To explicitly convert a value to boolean, use the (bool) or (boolean) casts. However, in most cases the cast is unnecessary, since a value will be **automatically converted** if an operator, function or control structure requires a boolean argument.
http://php.net/manual/en/language.types.boolean.php#language.types.boolean.casting

So event when `$var` is an array `if ($var)` is evaluated as `if var is true` according to PHP documentation. `if ($var)` is equivalent to `if ($var == true)` - non-strict comparison to a boolean.

---
> The more important question is: Why the array check now? What happens if it is not an array?

If `$range` is not an array but `1` or `true` then `$start` and `$end` are assigned `null` values although `false` values are expected in such scenario - take a look [here](http://sandbox.onlinephpfunctions.com/code/6623b50ebc865963924eb4fa8730c38bd2d4d922) again and compare it with the [correct](http://sandbox.onlinephpfunctions.com/code/f1307e1bdb883a91cb6a741342cc20a535ba8e66) implementation please.

Comment (dereuromark) [reply lib/Cake/Network/CakeResponse.php]: Then `if ($range && is_array($range)) {` is the correct code here.

Comment (bancer) [reply lib/Cake/Network/CakeResponse.php]: Okay, but when you claimed that `in PHP it doesnt mean true` you de-railed the discussion to PHP level yourself.

Comment (ADmad) [reply lib/Cake/Network/CakeResponse.php]: TLDR:  Don't make any changes that can potentially change the way code behaves. 2.x is **done**.

**Review by dereuromark — APPROVED** [review]
